---
title: "Genetic Constraints TagSeq"
author: "Rachel Wright"
date: "6/23/2021"
output: html_document
---

```{r Load libraries, echo = F}
library(DESeq2) # for differential gene expression analysis
library(tidyverse) # for wrangling and plotting
library(VennDiagram) # for venn diagram
library(pheatmap) # for pretty heatmaps
library(arrayQualityMetrics) # for sample outlier analysis
```

Sequencing stats

```{r sequencing stats}
countPreTrim <- read.delim("../countsPreTrim.txt", header=F)
# remove the first two rows of description
countPreTrim <- countPreTrim[-c(1,2),]
countPreTrim$V1 <- gsub(".fq", "", countPreTrim$V1)
names(countPreTrim)[2] <- "preTrim"

countPostTrim <- read.delim("../countsTrimmed.txt", header=F)
countPostTrim <- countPostTrim[-c(1,2),]
countPostTrim$V1 <- gsub(".fq.trim", "", countPostTrim$V1)
names(countPostTrim)[2] <- "postTrim"

mappedCounts <- read.delim("../AlignmentRate.txt", header=F)
mapped <- as.numeric(substr(mappedCounts$V1,1,5))

allStats <- merge(countPreTrim,countPostTrim,by="V1")
allStats <- cbind(allStats,mapped)

allStats %>% summarise(meanPreTrim = mean(preTrim),
                       meanPostTrim = mean(postTrim),
                       meanTrimPercRemaining = mean(postTrim/preTrim),
                       meanMappedPerc = mean(mapped))
```

# Load and format expression data

```{r Load expression data}
countdata0 <- read.delim("../allcounts_gencon_june21.txt")
dim(countdata0)

# look at the corners
countdata0[1:4,1:3]
countdata0[(nrow(countdata0)-4):nrow(countdata0),1:3]
countdata0[(nrow(countdata0)-4):nrow(countdata0),(length(countdata0)-4):length(countdata0)]

# get rid of the column called "X.1" that contains NAs
countdata0$X.1 <- NULL
```

Look at the corners again to make sure the file looks OK

```{r check expression data}
countdata0[1:4,1:3]
countdata0[(nrow(countdata0)-4):nrow(countdata0),1:3]
countdata0[(nrow(countdata0)-4):nrow(countdata0),(length(countdata0)-4):length(countdata0)]
countdata0[1:4,(length(countdata0)-4):length(countdata0)]
```

Separate the coral host and algal symbiont genes
Only run ONE of the lines below depending on which you want to analyze

```{r separate host and symbiont}
countdata <- countdata0 %>% filter(grepl("Amillepora", X)) %>% column_to_rownames('X') # for CORAL genes
# countdata <- countdata0 %>% filter(!grepl("Amillepora", X)) %>% column_to_rownames('X') # for SYMBIONT genes
```


Make column names nicer

```{r tidy names}
tank <- as.factor(substr(names(countdata),2,3))
genet <- as.factor(substr(names(countdata),5,6))
names(countdata) <- paste("t",tank,"_g",genet,sep="")
head(names(countdata))
```


We identified clones in the data set. Fix them.
11 and 12 = 60 (Rib)
24 and 32 = 61 (Pandora)
18 and 20 = 62 (Rib)
Add "a" and "b" to indicate the replicates

```{r fix clones}
names(countdata) <- gsub("g11", "g60a", names(countdata))
names(countdata) <- gsub("g12", "g60b", names(countdata))
names(countdata) <- gsub("g24", "g61a", names(countdata))
names(countdata) <- gsub("g32", "g61b", names(countdata))
names(countdata) <- gsub("g18", "g62a", names(countdata))
names(countdata) <- gsub("g20", "g62b", names(countdata))
# Now pad the other names with "a" to indicate the single replicate for that genet/treat
names(countdata) <- str_pad(names(countdata), 8, 
                            side = "right", pad = "a")
head(names(countdata))
```

# Load and format the trait data

```{r}
load("../../AllTraits_May2018/traits/allTraitData_Survival.RData")

# summarize by tank and genet (average biological replicates)
d_cond_data <- d %>% ungroup() %>% 
  dplyr::select(tank,geno,tank,reef,treat) %>% 
  rename(genet = geno) %>%
  mutate(sam = as.factor(paste("t",tank,"_g",genet,"a",sep=""))) %>%
  distinct(sam, .keep_all=T)

# make rows with the other replicate
dup_rows <- d_cond_data %>% filter (genet %in% c("60","61","62"))
dup_rows$sam <- gsub("a","b",dup_rows$sam)

# add to the other data frame
conds <- rbind(d_cond_data,dup_rows)

conds <- conds[conds$sam %in% names(countdata),]

# sort to match the order in the counts file
conds <- conds[match(names(countdata), conds$sam),]
table(conds$sam == names(countdata))

# summarize samples
conds %>% group_by(treat) %>% summarize(n = n())
```

Construct data object for DESeq2

```{r dds}
dds <- DESeqDataSetFromMatrix(
  countData = countdata,
  colData = conds,
  design = ~ treat + genet)
```

Set the base mean minimum to filter lowly abundant genes

```{r filter genes}
means <- apply(countdata,1,mean)
table(means>3)

means3 <- names(means[means>3])

countFilt <- countdata[row.names(countdata) %in% means3,]

totalCountsFilt <- colSums(countFilt)
min(totalCountsFilt) # 64887
max(totalCountsFilt) # 507190
mean(totalCountsFilt) # 224031
```

Important: check that the sample order in the count file exactly matches the sample order in the conditions file

```{r check sample order}
table(names(countFilt) == as.vector(conds$sam))
```

Re-construct data object for DESeq2

```{r dds filtered}
ddsFilt <- DESeqDataSetFromMatrix(
  countData = countFilt,
  colData = conds,
  design = ~ treat + genet)
```

Perform DESeq2 analysis

```{r deds}
deds <- DESeq(ddsFilt)
```

Look at the results

```{r results - bacteria}
resBac <- results(deds, independentFiltering = F, contrast = c("treat","b","c"))
summary(resBac, alpha = 0.05)
```

```{r results - combined}
resComb <- results(deds, independentFiltering = F, contrast = c("treat","a","c"))
summary(resComb, alpha = 0.05)
```

```{r results - heat}
resHeat <- results(deds, independentFiltering = F, contrast = c("treat","h","c"))
summary(resHeat, alpha = 0.05)
```

```{r results - PCo2}
resPco2 <- results(deds, independentFiltering = F, contrast = c("treat","p","c"))
summary(resPco2, alpha = 0.05)
```


